File: loan-sense-frontend/server/ml/predict.py
import joblib
import numpy as np

class LoanRiskPredictor:
    def __init__(self):
        self.model = joblib.load('xgboost_model.joblib')
        self.label_encoders = joblib.load('label_encoders.joblib')
        
        self.features = [
            'Age', 'Income', 'LoanAmount', 'CreditScore', 'MonthsEmployed',
            'DTIRatio', 'Education', 'EmploymentType', 'MaritalStatus',
            'HasMortgage', 'HasDependents', 'LoanPurpose', 'HasCoSigner'
        ]
        
    def preprocess_input(self, data):
        # Convert categorical variables
        for column, encoder in self.label_encoders.items():
            if column in data:
                data[column] = encoder.transform([data[column]])[0]
        
        # Convert boolean variables to int
        boolean_columns = ['HasMortgage', 'HasDependents', 'HasCoSigner']
        for column in boolean_columns:
            if column in data:
                data[column] = 1 if data[column] == 'Yes' else 0
        
        # Create feature array in correct order
        features = np.zeros(len(self.features))
        for i, feature in enumerate(self.features):
            features[i] = data.get(feature, 0)
            
        return features.reshape(1, -1)
    
    def predict_risk(self, data):
        # Preprocess input data
        features = self.preprocess_input(data)
        
        # Get probability of default
        default_prob = self.model.predict_proba(features)[0][1]
        
        # Convert to risk rating (0-100)
        risk_rating = int(default_prob * 100)
        
        # Get risk category
        if risk_rating < 20:
            category = "Very Low Risk"
        elif risk_rating < 40:
            category = "Low Risk"
        elif risk_rating < 60:
            category = "Moderate Risk"
        elif risk_rating < 80:
            category = "High Risk"
        else:
            category = "Very High Risk"
            
        return {
            "risk_rating": risk_rating,
            "risk_category": category,
            "default_probability": float(default_prob)
        } 

File: loan-sense-frontend/server/ml/predict.js
// Mock risk prediction function
function predict_risk(loanData) {
    // Simple risk calculation based on key factors
    let riskScore = 0;
    
    // Credit score impact (higher is better)
    if (loanData.creditScore >= 800) riskScore += 20;
    else if (loanData.creditScore >= 700) riskScore += 15;
    else if (loanData.creditScore >= 600) riskScore += 10;
    else riskScore += 5;
    
    // Income to loan ratio (higher is better)
    const incomeToLoanRatio = loanData.income / loanData.loanAmount;
    if (incomeToLoanRatio >= 0.5) riskScore += 20;
    else if (incomeToLoanRatio >= 0.3) riskScore += 15;
    else if (incomeToLoanRatio >= 0.2) riskScore += 10;
    else riskScore += 5;
    
    // DTI ratio impact (lower is better)
    if (loanData.dtiRatio <= 20) riskScore += 20;
    else if (loanData.dtiRatio <= 30) riskScore += 15;
    else if (loanData.dtiRatio <= 40) riskScore += 10;
    else riskScore += 5;
    
    // Employment stability
    if (loanData.monthsEmployed >= 60) riskScore += 20;
    else if (loanData.monthsEmployed >= 36) riskScore += 15;
    else if (loanData.monthsEmployed >= 12) riskScore += 10;
    else riskScore += 5;
    
    // Additional factors
    if (loanData.hasCoSigner === true) riskScore += 10;
    if (loanData.hasMortgage === false) riskScore += 5;
    if (loanData.hasDependents === false) riskScore += 5;
    
    // Calculate final risk rating (0-100)
    const riskRating = Math.min(100, Math.max(0, riskScore));
    
    // Determine risk category
    let riskCategory;
    if (riskRating >= 80) riskCategory = 'Low Risk';
    else if (riskRating >= 60) riskCategory = 'Medium Risk';
    else riskCategory = 'High Risk';
    
    return {
        risk_rating: riskRating,
        risk_category: riskCategory
    };
}

module.exports = {
    predict_risk
}; 